docs/blog/posts/2024-12-11-ultralytics-attack-analysis.md

```diff
@@ -0,0 +1,68 @@
+---
+title: "Supply-chain attack analysis: Ultralytics"
+description: Analysis of a package targeted by a supply-chain attack to the build and release process
+authors:
+  - sethmlarson
+date: 2024-12-11
+tags:
+  - security
+---
+
+Last week, the Python project “[ultralytics](https://pypi.org/project/ultralytics/)” suffered a supply-chain attack through a compromise of the projects’ GitHub Actions workflows and subsequently its PyPI API token. No security flaw in PyPI was used to execute this attack. Versions 8.3.41, 8.3.42, 8.3.45, and 8.3.46 were affected and have been removed from PyPI.
+
+<!-- more -->
+
+The attack highlights the importance of securing software forges and the build and publish workflows for open source projects. A [complete set of details is available](https://blog.yossarian.net/2024/12/06/zizmor-ultralytics-injection#conclusions) thanks to the analysis of William Woodruff from Trail of Bits.
+
+## Visibility from Attestations and Trusted Publishers
+
+Despite the success of the attack, many things went right from PyPI’s perspective, especially the ability to audit the attack while it was occurring and after the fact. Because the Ultralytics project was using Trusted Publishing and the [PyPA’s publishing GitHub Action](https://github.com/pypa/gh-action-pypi-publish): PyPI staff, volunteers, and security researchers were able to dig into how maliciously injected software was able to make its way into the package.
+
+From looking at the [Sigstore transparency logs](https://search.sigstore.dev/?logIndex=153589716) and the PyPI provenance attestations, it was clear that the first set of injected packages were published through the existing GitHub Actions workflow, not by an API token. This considerably reduced the scope of the attack: either the malicious code was inside the source repository or was injected during the build phase. Later investigation showed that the [attack targeted the GitHub Actions cache](https://blog.yossarian.net/2024/12/06/zizmor-ultralytics-injection#conclusions) which was used during the build phase.
+
+The second round of malicious releases came from the attacker using an unrevoked PyPI API token that was still available to the GitHub Actions workflow, potentially a hold-over from before the project adopted Trusted Publishing. This was detectable because there were no corresponding source repository activity or or PyPI publish attestations for the second round of releases.
```

Comment:
```suggestion
The second round of malicious releases came from the attacker using an unrevoked PyPI API token that was still available to the GitHub Actions workflow, potentially a hold-over from before the project adopted Trusted Publishing. This was detectable because there were no corresponding source repository activity or PyPI publish attestations for the second round of releases.
```